### ISSUE TITLE
ERROR: from element /GstPipeline:pipeline0/GstV4l2Src:v4l2src0: Internal data stream error.

### ISSUE THREAD

**Issue by pietrushnic** [elliotkillick/qubes-video-companion#21]
TBH I have no idea if I'm doing it correctly. My procedure:

1. Compile deb package
2. Install in `work` VM (here I want to use C922) and in `sys-usb` VM (camera is connected to this VM) package `qubes-video-companion_1.0.0-1_all.deb`
3. Change `qubes-video-companion` so it points to `sys-usb` instead of `dom0`
4. First call `DEBUG=1 qubes-video-companion webcam` asked me for permission and I answered `YES`, then I had to point to `sys-usb`

Results looks as follows:

```
work% DEBUG=1 qubes-video-companion webcam
+ set -E
+ trap exit ERR
+ video_source=webcam
+ case "$video_source" in
+ qvc_service=qvc.Webcam
+ qvc_lock_file=/run/lock/qubes-video-companion
+ '[' -f /run/lock/qubes-video-companion ']'
+ trap exit_clean EXIT
+ sudo touch /run/lock/qubes-video-companion
+ /usr/share/qubes-video-companion/video/setup.sh
+ set -E
+ trap exit ERR
+ source /usr/share/qubes-video-companion/video/common.sh
++ '[' 1 == 1 ']'
++ set -x
++ set -E
++ trap exit ERR
+ test_v4l2loopback
+ PATH=/usr/local/sbin:/usr/sbin:/sbin:/usr/local/bin:/usr/bin:/bin:/usr/local/games:/usr/games
+ modinfo v4l2loopback
+ sudo modprobe v4l2loopback 'card_label=Qubes Video Companion' exclusive_caps=1
+ sudo setfacl -m user:user:rw /dev/video0
+ qrexec-client-vm sys-usb qvc.Webcam /usr/share/qubes-video-companion/video/receiver.sh --buffer-size=0 --filter-escape-chars-stderr
+ set -E
+ trap exit ERR
+ read -r untrusted_width untrusted_height untrusted_fps
Starting webcam stream at 640x480 30 FPS...
++ sanitize_cap 640
++ local cap_value=640
++ echo 640
++ tr -cd 0-9
+ width=640
++ sanitize_cap 480
++ local cap_value=480
++ echo 480
++ tr -cd 0-9
+ height=480
++ sanitize_cap 30
++ local cap_value=30
++ echo 30
++ tr -cd 0-9
+ fps=30
+ '[' 640 ']'
+ echo 'Receiving video stream at 640x480 30 FPS...'
Receiving video stream at 640x480 30 FPS...
+ frame_rate=30/1
+ gst-launch-1.0 fdsrc '!' queue '!' capsfilter caps=video/x-raw,width=640,height=480,framerate=30/1,format=I420,colorimetry=2:4:7:1,chroma-site=none,interlace-mode=progressive,pixel-aspect-ratio=1/1,max-framerate=30/1,views=1 '!' rawvideoparse use-sink-caps=true '!' v4l2sink device=/dev/video0 sync=false
ERROR: from element /GstPipeline:pipeline0/GstV4l2Src:v4l2src0: Internal data stream error.
Additional debug info:
gstbasesrc.c(3055): gst_base_src_loop (): /GstPipeline:pipeline0/GstV4l2Src:v4l2src0:
streaming stopped, reason not-negotiated (-4)
++ exit
++ exit
+ exit_clean
+ exit_code=141
+ /usr/share/qubes-video-companion/video/destroy.sh
+ set -E
+ trap exit ERR
+ source /usr/share/qubes-video-companion/video/common.sh
++ '[' 1 == 1 ']'
++ set -x
++ set -E
++ trap exit ERR
+ test_v4l2loopback
+ PATH=/usr/local/sbin:/usr/sbin:/sbin:/usr/local/bin:/usr/bin:/bin:/usr/local/games:/usr/games
+ modinfo v4l2loopback
+ sudo modprobe -r v4l2loopback videodev
+ sudo rm -f /run/lock/qubes-video-companion
+ '[' webcam == webcam ']'
+ '[' 141 == 141 ']'
+ echo 'The webcam device is in use! Please stop any instance of Qubes Video Companion running on another qube.'
The webcam device is in use! Please stop any instance of Qubes Video Companion running on another qube.
+ exit 1
```

Testing procedure have to be described better since I have no idea if I'm doing it correctly.

Personally I would be glad to see this on Qubes OS minisummit: https://twitter.com/3mdeb_com/status/1412759830062899210

**Comment by elliotkillick** [comment]
`qubes-video-companion` is supposed to point to `dom0` because that is where the qrexec policies are stored in `/etc/qubes-rpc/policy`. It's within that directory that you should create a new file called `qvc.Webcam` and write [this](https://github.com/elliotkillick/qubes-video-companion/blob/master/qubes-rpc/policies/qvc.Webcam) inside of it. Perform the same process for `qvc.ScreenShare`. At this point, your webcam in `sys-usb` and screen sharing between AppVMs should be working.

Note that if you want to do screen sharing from `dom0` then you will have to install the RPM package in `dom0` (this will also install the qrexec policies for you). The package can be created by running `./create-rpm.sh dom0` in a Fedora 25 VM (same version as `dom0`) using [this](https://github.com/elliotkillick/qubes-video-companion/blob/master/build/create-rpm.sh) script.

Let me know if you're able to get this working! Interesting minisummit you have going on there, I hope Qubes Video Companion makes it in.

**Comment by pietrushnic** [comment]
@elliotkillick I'm trying to preapre presentation for minisummit, but can't make qubes-video-comapnion work. What I have is `work` and `sys-usb` VM based on debian-10. `sys-usb` have Logitech C922 connected and I would like to use camera inside `work` VM.

What I tried is exactly what you explained. I built deb package using `build/create-deb.sh`, which works fine except fact that path for deb package advertised in final comment is different:

```
Run the following command one directory up from the qubes-video-companion directory to install the package: sudo apt install ./qubes-video-companion_1.0.0-1_all.deb
work% sudo apt install ./qubes-video-companion_1.0.0-1_all.deb
Reading package lists... Done
E: Unsupported file ./qubes-video-companion_1.0.0-1_all.deb given on commandline
```

It should be `../qubes-video-companion_1.0.0-1_all.deb`. Then installation end up with following message:

```
N: Download is performed unsandboxed as root as file '/home/user/Downloads/src/qubes/qubes-video-companion_1.0.0-1_all.deb' couldn't be accessed by user '_apt'. - pkgAcquire::Run (13: Permission denied)
```

After that I created `qvc.Webcam` and `qvc.ScreenShare` in dom0 as well as installed package in `sys-usb`. All that doesn't help since when I call `DEBUG=1 qubes-video-companion webcam` and point to `sys-usb` I'm getting exactly the same GStreamer error as before:

```
+ set -E
+ trap exit ERR
+ video_source=webcam
+ case "$video_source" in
+ qvc_service=qvc.Webcam
+ qvc_lock_file=/run/lock/qubes-video-companion
+ '[' -f /run/lock/qubes-video-companion ']'
+ trap exit_clean EXIT
+ sudo touch /run/lock/qubes-video-companion
+ /usr/share/qubes-video-companion/video/setup.sh
+ set -E
+ trap exit ERR
+ source /usr/share/qubes-video-companion/video/common.sh
++ '[' 1 == 1 ']'
++ set -x
++ set -E
++ trap exit ERR
+ test_v4l2loopback
+ PATH=/usr/local/sbin:/usr/sbin:/sbin:/usr/local/bin:/usr/bin:/bin:/usr/local/games:/usr/games
+ modinfo v4l2loopback
+ sudo modprobe v4l2loopback 'card_label=Qubes Video Companion' exclusive_caps=1
+ sudo setfacl -m user:user:rw /dev/video0
+ qrexec-client-vm dom0 qvc.Webcam /usr/share/qubes-video-companion/video/receiver.sh --buffer-size=0 --filter-escape-chars-stderr
+ set -E
+ trap exit ERR
+ read -r untrusted_width untrusted_height untrusted_fps
Starting webcam stream at 640x480 30 FPS...
++ sanitize_cap 640
++ local cap_value=640
++ echo 640
++ tr -cd 0-9
+ width=640
++ sanitize_cap 480
++ local cap_value=480
++ echo 480
++ tr -cd 0-9
+ height=480
++ sanitize_cap 30
++ local cap_value=30
++ tr -cd 0-9
++ echo 30
+ fps=30
+ '[' 640 ']'
+ echo 'Receiving video stream at 640x480 30 FPS...'
Receiving video stream at 640x480 30 FPS...
+ frame_rate=30/1
+ gst-launch-1.0 fdsrc '!' queue '!' capsfilter caps=video/x-raw,width=640,height=480,framerate=30/1,format=I420,colorimetry=2:4:7:1,chroma-site=none,interlace-mode=progressive,pixel-aspect-ratio=1/1,max-framerate=30/1,views=1 '!' rawvideoparse use-sink-caps=true '!' v4l2sink device=/dev/video0 sync=false
ERROR: from element /GstPipeline:pipeline0/GstV4l2Src:v4l2src0: Internal data stream error.
Additional debug info:
gstbasesrc.c(3055): gst_base_src_loop (): /GstPipeline:pipeline0/GstV4l2Src:v4l2src0:
streaming stopped, reason not-negotiated (-4)
+ exit_clean
+ exit_code=0
+ /usr/share/qubes-video-companion/video/destroy.sh
+ set -E
+ trap exit ERR
+ source /usr/share/qubes-video-companion/video/common.sh
++ '[' 1 == 1 ']'
++ set -x
++ set -E
++ trap exit ERR
+ test_v4l2loopback
+ PATH=/usr/local/sbin:/usr/sbin:/sbin:/usr/local/bin:/usr/bin:/bin:/usr/local/games:/usr/games
+ modinfo v4l2loopback
+ sudo modprobe -r v4l2loopback videodev
+ sudo rm -f /run/lock/qubes-video-companion
+ '[' webcam == webcam ']'
+ '[' 0 == 141 ']'
+ exit 0
```

From Makefile I see that `install-dom0` target put in dom0 quite more then just qvc files, is `qubes-video-companion` installation needed in `dom0` ? Any idea what I'm doing wrong here?

**Comment by elliotkillick** [comment]
@pietrushnic Sorry I couldn't get back to you earlier, I was away for vacation and then had some things I needed to tend to.

Can you please test to see if screen sharing is working (e.g. from `sys-usb` to another AppVM)? If that's working but not the webcam functionality then I think I know the issue.

No, the `qubes-video-companion` command should not exist in `dom0`.

**Comment by elliotkillick** [comment]
I think you're working with the old `v1.0.4` branch, please make the changes here to the sending VM `qvm.Webcam` script: https://github.com/elliotkillick/qubes-video-companion/commit/26f91b6b043e7c981531c1a29426fa0d84a46664

**Comment by pietrushnic** [comment]
@elliotkillick I'm using most recent code from master branch in `debian-10` vms. Compilation worked without problems, the remaingin things I had to do to move forward were:
```
/usr/share/qubes-video-companion/scripts/v4l2loopback/install.sh
sudo apt install v4l-utils #both in recipient as well as in sys-usb
```

 It seem that I have all in place, but both webcam and screenshare commands complain as follows:

```
[2:28:51] user:qubes-video-companion git:(d89bd8b8d0e4) $ qubes-video-companion webcam      
Starting webcam stream at 640x480 30 FPS...
Receiving video stream at 640x480 30 FPS...
ERROR: from element /GstPipeline:pipeline0/GstV4l2Src:v4l2src0: Internal data stream error.
Additional debug info:
gstbasesrc.c(3055): gst_base_src_loop (): /GstPipeline:pipeline0/GstV4l2Src:v4l2src0:
streaming stopped, reason not-negotiated (-4)
The webcam device is in use! Please stop any instance of Qubes Video Companion running on another qube.
[2:28:55] user:qubes-video-companion git:(d89bd8b8d0e4) $ qubes-video-companion screenshare
Starting screen sharing at 1920x1200 30 FPS...
Receiving video stream at 1920x1200 30 FPS...
WARNING: erroneous pipeline: could not link queue0 to videoconvert0, queue0 can't handle caps video/x-raw, width=(int)1920, height=(int)1200, framerate=(fraction)30/1, format=(string)BGRx
```

Also tried your modification from previous post: https://github.com/elliotkillick/qubes-video-companion/commit/26f91b6b043e7c981531c1a29426fa0d84a46664
But that doesn't change anything.

**Comment by pietrushnic** [comment]
Ok, I removed checkout of release tag and from master I managed to make it work. Testing it. Thanks for the support. I will get back with the feedback.

**Comment by elliotkillick** [comment]
> Ok, I removed checkout of release tag and from master I managed to make it work. Testing it. Thanks for the support. I will get back with the feedback.

Great, I just released version 2.0.0 with all the latest and greatest changes so checking out the tag should work now! It's important to security because the build scripts verify the tag against my PGP key.

**Comment by pietrushnic** [comment]
I think we got success presented at Qubes OS mini-summit. Issue can be closed.
